File: Go/0055-JumpGame/Solution.go
package lc55

import lc45 "gitlab.com/euchangxian/leetcode/Go/0045-JumpGame"

// lc45.Jump returns the minimum number of jumps required to reach the last index. 0 represents impossible to reach the last index
// As such, we can reduce the optimization problem: find the minimum number of jumps to reach the last index, to a
// decision problem: is it possible to reach the last index in k jumps.
func canJump(nums []int) bool {
	return lc45.Jump(nums) > 0
}

File: C++/0055-JumpGame/Solution.cpp
#include <vector>

class Solution {
 public:
  bool canJump(std::vector<int>& nums) {
    int n = nums.size();
    int jumps = 1;
    int furthest = nums[0];  // Represents the furthest index reachable
    for (int i = 1; i < n; ++i) {
      if (i > furthest) {
        return false;
      }

      if (i + nums[i] > furthest) {
        furthest = i + nums[i];
      }

      if (furthest >= n - 1) {
        break;
      }
    }
    return true;
  }
};
